### ISSUE TITLE
Bundler installs unsuitable gem versions on Windows

### ISSUE THREAD

**Issue by leoarnold** [rubygems/bundler#5992]
**Environments:** (all possible combinations were tested)
* Windows 10 or Windows Server 2012
* Ruby:
   * ruby 2.1.9p490 (2016-03-30 revision 54437) [x64-mingw32]
   * ruby 2.1.9p490 (2016-03-30 revision 54437) [i386-mingw32]
   * ruby 2.0.0p648 (2015-12-16) [x64-mingw32]
   * ruby 2.0.0p648 (2015-12-16) [i386-mingw32]
   * ruby 1.9.3p551 (2014-11-13) [i386-mingw32]
* Gem 2.6.13
* Bundler 1.15.4
* Gemfile:

```ruby
source 'https://rubygems.org'

gem 'nokogiri'
```

**Problem:** Bundler installs a version of Nokogiri which requires `< 2.5, >= 2.2` even though Bundler is running on Ruby `< 2.2`.

```
Fetching nokogiri 1.8.0 (x64-mingw32) # respectivly (x86-mingw32)
Installing nokogiri 1.8.0 (x64-mingw32)
Gem::InstallError: nokogiri requires Ruby version < 2.5, >= 2.2.
An error occurred while installing nokogiri (1.8.0), and Bundler cannot
continue.
Make sure that `gem install nokogiri -v '1.8.0'` succeeds before bundling.
In Gemfile:
  nokogiri
Command exited with code 5
```

**Suspicion:** Bundler chooses the `mingw32` version, but uses the metadata from the mainline gem. In the case of Nokogiri 1.8.0, those differ:
* [nokogiri 1.8.0](https://rubygems.org/gems/nokogiri/versions/1.8.0) requires Ruby `>= 2.1.0`
* [nokogiri 1.8.0 (x86-mingw32)](https://rubygems.org/gems/nokogiri/versions/1.8.0-x86-mingw32) requires Ruby `< 2.5, >= 2.2 `
* [nokogiri 1.8.0 (x64-mingw32)](https://rubygems.org/gems/nokogiri/versions/1.8.0-x64-mingw32) requires Ruby `< 2.5, >= 2.2 `

**Caveat:** Using the `nokogiri` example, the bug cannot be reproduced on Ruby `>= 2.2.0`. This does *NOT* mean that the bug is not there - it's just the case where the differing Ruby version requirements currently overlap.

**Comment by segiddins** [comment]
Please share the output of `bundle env`

**Comment by segiddins** [comment]
If your suspicion is correct, it will be fixed by enabling the `specific_platform` setting, which will default to `true` on Bundler 2.0.

**Comment by leoarnold** [comment]
# bundle env
## Environment
```
Bundler   1.15.4
Rubygems  2.6.13
Ruby      2.1.9p490 (2016-03-30 revision 54437) [i386-mingw32]
Git       2.13.0.windows.1
Platform  x86-mingw32
OpenSSL   OpenSSL 1.0.1l 15 Jan 2015
```
## Gemfile
### Gemfile
```ruby
source 'https://rubygems.org'
gem 'nokogiri'
```
### Gemfile.lock
```
<No C:/[...]/Gemfile.lock found>
```

**Comment by segiddins** [comment]
Can you try `bundle config --local specific_platform true && bundle install` ?

**Comment by leoarnold** [comment]
Does not change a thing:
```
ruby -v
ruby 2.1.9p490 (2016-03-30 revision 54437) [i386-mingw32]
gem -v
2.6.13
bundle -v
Bundler version 1.15.4
bundle config --local specific_platform true
You are replacing the current local value of specific_platform, which is currently nil
bundle env
```

## Environment
```
Bundler   1.15.4
Rubygems  2.6.13
Ruby      2.1.9p490 (2016-03-30 revision 54437) [i386-mingw32]
Git       2.13.0.windows.1
Platform  x86-mingw32
OpenSSL   OpenSSL 1.0.1l 15 Jan 2015
```
## Bundler settings
```
specific_platform
  Set for your local app (C:/[...]/.bundle/config): "true"
```
## Gemfile
### Gemfile
```ruby
source 'https://rubygems.org'
gem 'nokogiri'
```
### Gemfile.lock
```
<No C:/[...]/Gemfile.lock found>
```

```
bundle install --path vendor/bundle
Fetching gem metadata from https://rubygems.org/.............
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using bundler 1.15.4
Fetching mini_portile2 2.2.0
Installing mini_portile2 2.2.0
Fetching nokogiri 1.8.0 (x86-mingw32)
Installing nokogiri 1.8.0 (x86-mingw32)
Gem::InstallError: nokogiri requires Ruby version < 2.5, >= 2.2.
An error occurred while installing nokogiri (1.8.0), and Bundler cannot
continue.
Make sure that `gem install nokogiri -v '1.8.0'` succeeds before bundling.
In Gemfile:
  nokogiri
Command exited with code 5
```

**Comment by segiddins** [comment]
What's the output of `bundle install -V`? I'm curious if you're not getting results from the compact index, which is the only index that contains the `required_ruby_version` info

**Comment by leoarnold** [comment]
## Environment
```
Bundler   1.15.4
Rubygems  2.6.13
Ruby      2.1.9p490 (2016-03-30 revision 54437) [i386-mingw32]
Git       2.13.0.windows.1
Platform  x86-mingw32
OpenSSL   OpenSSL 1.0.1l 15 Jan 2015
```
## Bundler settings
```
specific_platform
  Set for your local app (C:/[...]/.bundle/config): "true"
```
## Gemfile
### Gemfile
```ruby
source 'https://rubygems.org'
gem 'nokogiri'
```
### Gemfile.lock
```
<No C:/[...]/Gemfile.lock found>
```

```
bundle install -V
Running `bundle install --verbose` with bundler 1.15.4
Found changes from the lockfile, re-resolving dependencies because the list of sources changed, the dependencies in your gemfile changed, you added a new platform to your gemfile
HTTP GET https://index.rubygems.org/versions
HTTP 200 OK https://index.rubygems.org/versions
Fetching gem metadata from https://rubygems.org/
Looking up gems ["nokogiri"]
HTTP GET https://index.rubygems.org/info/nokogiri
HTTP 200 OK https://index.rubygems.org/info/nokogiri
Looking up gems ["hoe", "racc", "rake-compiler", "rexical", "tenderlove-frex", "weakling", "mini_portile", "mini_portile2", "pkg-config"]
HTTP GET https://index.rubygems.org/info/hoe
HTTP GET https://index.rubygems.org/info/racc
HTTP GET https://index.rubygems.org/info/rake-compiler
HTTP GET https://index.rubygems.org/info/rexical
HTTP GET https://index.rubygems.org/info/tenderlove-frex
HTTP GET https://index.rubygems.org/info/weakling
HTTP GET https://index.rubygems.org/info/mini_portile
HTTP GET https://index.rubygems.org/info/mini_portile2
HTTP GET https://index.rubygems.org/info/pkg-config
HTTP 200 OK https://index.rubygems.org/info/hoe
HTTP 200 OK https://index.rubygems.org/info/mini_portile
HTTP 200 OK https://index.rubygems.org/info/tenderlove-frex
HTTP 200 OK https://index.rubygems.org/info/mini_portile2
HTTP 200 OK https://index.rubygems.org/info/pkg-config
HTTP 200 OK https://index.rubygems.org/info/racc
HTTP 200 OK https://index.rubygems.org/info/rexical
HTTP 200 OK https://index.rubygems.org/info/rake-compiler
HTTP 200 OK https://index.rubygems.org/info/weakling
Looking up gems ["rake", "rubyforge", "RubyInline", "gemcutter"]
HTTP GET https://index.rubygems.org/info/rake
HTTP GET https://index.rubygems.org/info/rubyforge
HTTP GET https://index.rubygems.org/info/RubyInline
HTTP GET https://index.rubygems.org/info/gemcutter
HTTP 200 OK https://index.rubygems.org/info/gemcutter
HTTP 200 OK https://index.rubygems.org/info/rake
HTTP 200 OK https://index.rubygems.org/info/RubyInline
HTTP 200 OK https://index.rubygems.org/info/rubyforge
Looking up gems ["json", "net-scp", "json_pure", "ZenTest"]
HTTP GET https://index.rubygems.org/info/json
HTTP GET https://index.rubygems.org/info/net-scp
HTTP GET https://index.rubygems.org/info/json_pure
HTTP GET https://index.rubygems.org/info/ZenTest
HTTP 200 OK https://index.rubygems.org/info/json_pure
HTTP 200 OK https://index.rubygems.org/info/json
HTTP 200 OK https://index.rubygems.org/info/net-scp
HTTP 200 OK https://index.rubygems.org/info/ZenTest
Looking up gems ["spruz", "net-ssh"]
HTTP GET https://index.rubygems.org/info/spruz
HTTP GET https://index.rubygems.org/info/net-ssh
HTTP 200 OK https://index.rubygems.org/info/net-ssh
HTTP 200 OK https://index.rubygems.org/info/spruz
Looking up gems ["echoe", "needle", "jruby-pageant", "bcrypt_pbkdf", "rbnacl", "rbnacl-libsodium"]
HTTP GET https://index.rubygems.org/info/echoe
HTTP GET https://index.rubygems.org/info/needle
HTTP GET https://index.rubygems.org/info/jruby-pageant
HTTP GET https://index.rubygems.org/info/bcrypt_pbkdf
HTTP GET https://index.rubygems.org/info/rbnacl
HTTP GET https://index.rubygems.org/info/rbnacl-libsodium
HTTP 200 OK https://index.rubygems.org/info/rbnacl-libsodium
HTTP 200 OK https://index.rubygems.org/info/bcrypt_pbkdf
HTTP 200 OK https://index.rubygems.org/info/echoe
HTTP 200 OK https://index.rubygems.org/info/jruby-pageant
HTTP 200 OK https://index.rubygems.org/info/needle
HTTP 200 OK https://index.rubygems.org/info/rbnacl
Looking up gems ["ffi", "highline", "rcov", "allison", "rdoc"]
HTTP GET https://index.rubygems.org/info/ffi
HTTP GET https://index.rubygems.org/info/highline
HTTP GET https://index.rubygems.org/info/rcov
HTTP GET https://index.rubygems.org/info/allison
HTTP GET https://index.rubygems.org/info/rdoc
HTTP 200 OK https://index.rubygems.org/info/highline
HTTP 200 OK https://index.rubygems.org/info/ffi
HTTP 200 OK https://index.rubygems.org/info/allison
HTTP 200 OK https://index.rubygems.org/info/rdoc
HTTP 200 OK https://index.rubygems.org/info/rcov
Looking up gems ["termios", "minitest", "bones"]
HTTP GET https://index.rubygems.org/info/termios
HTTP GET https://index.rubygems.org/info/minitest
HTTP GET https://index.rubygems.org/info/bones
HTTP 200 OK https://index.rubygems.org/info/termios
HTTP 200 OK https://index.rubygems.org/info/minitest
HTTP 200 OK https://index.rubygems.org/info/bones
Looking up gems ["little-plugger", "loquacious", "bones-extras", "bones-git", "rspec"]
HTTP GET https://index.rubygems.org/info/little-plugger
HTTP GET https://index.rubygems.org/info/loquacious
HTTP GET https://index.rubygems.org/info/bones-extras
HTTP GET https://index.rubygems.org/info/rspec
HTTP GET https://index.rubygems.org/info/bones-git
HTTP 200 OK https://index.rubygems.org/info/bones-extras
HTTP 200 OK https://index.rubygems.org/info/rspec
HTTP 200 OK https://index.rubygems.org/info/loquacious
HTTP 200 OK https://index.rubygems.org/info/little-plugger
HTTP 200 OK https://index.rubygems.org/info/bones-git
Looking up gems ["bones-rcov", "bones-rspec", "bones-rubyforge", "bones-zentest", "cucumber", "diff-lcs", "spicycode-rcov", "syntax", "rspec-core", "rspec-expectations", "rspec-mocks", "git"]
HTTP GET https://index.rubygems.org/info/bones-rcov
HTTP GET https://index.rubygems.org/info/bones-rspec
HTTP GET https://index.rubygems.org/info/bones-rubyforge
HTTP GET https://index.rubygems.org/info/bones-zentest
HTTP GET https://index.rubygems.org/info/cucumber
HTTP GET https://index.rubygems.org/info/spicycode-rcov
HTTP GET https://index.rubygems.org/info/syntax
HTTP GET https://index.rubygems.org/info/rspec-expectations
HTTP GET https://index.rubygems.org/info/rspec-mocks
HTTP GET https://index.rubygems.org/info/git
HTTP GET https://index.rubygems.org/info/diff-lcs
HTTP GET https://index.rubygems.org/info/rspec-core
HTTP 200 OK https://index.rubygems.org/info/diff-lcs
HTTP 200 OK https://index.rubygems.org/info/rspec-mocks
HTTP 200 OK https://index.rubygems.org/info/cucumber
HTTP 200 OK https://index.rubygems.org/info/bones-rubyforge
HTTP 200 OK https://index.rubygems.org/info/git
HTTP 200 OK https://index.rubygems.org/info/syntax
HTTP 200 OK https://index.rubygems.org/info/bones-zentest
HTTP 200 OK https://index.rubygems.org/info/bones-rspec
HTTP 200 OK https://index.rubygems.org/info/bones-rcov
HTTP 200 OK https://index.rubygems.org/info/rspec-core
HTTP 200 OK https://index.rubygems.org/info/rspec-expectations
HTTP 200 OK https://index.rubygems.org/info/spicycode-rcov
Looking up gems ["rspec-support", "builder", "polyglot", "term-ansicolor", "treetop", "gherkin", "multi_json", "multi_test", "cucumber-core", "gherkin3", "cucumber-wire", "event-bus", "cucumber-expressions"]
HTTP GET https://index.rubygems.org/info/builder
HTTP GET https://index.rubygems.org/info/polyglot
HTTP GET https://index.rubygems.org/info/term-ansicolor
HTTP GET https://index.rubygems.org/info/treetop
HTTP GET https://index.rubygems.org/info/gherkin
HTTP GET https://index.rubygems.org/info/multi_json
HTTP GET https://index.rubygems.org/info/multi_test
HTTP GET https://index.rubygems.org/info/cucumber-core
HTTP GET https://index.rubygems.org/info/gherkin3
HTTP GET https://index.rubygems.org/info/cucumber-wire
HTTP GET https://index.rubygems.org/info/event-bus
HTTP GET https://index.rubygems.org/info/cucumber-expressions
HTTP GET https://index.rubygems.org/info/rspec-support
HTTP 200 OK https://index.rubygems.org/info/cucumber-expressions
HTTP 200 OK https://index.rubygems.org/info/event-bus
HTTP 200 OK https://index.rubygems.org/info/gherkin3
HTTP 200 OK https://index.rubygems.org/info/rspec-support
HTTP 200 OK https://index.rubygems.org/info/cucumber-wire
HTTP 200 OK https://index.rubygems.org/info/multi_test
HTTP 200 OK https://index.rubygems.org/info/term-ansicolor
HTTP 200 OK https://index.rubygems.org/info/cucumber-core
HTTP 200 OK https://index.rubygems.org/info/multi_json
HTTP 200 OK https://index.rubygems.org/info/treetop
HTTP 200 OK https://index.rubygems.org/info/polyglot
HTTP 200 OK https://index.rubygems.org/info/builder
HTTP 200 OK https://index.rubygems.org/info/gherkin
Looking up gems ["tins", "backports", "cucumber-tag_expressions", "facets", "trollop"]
HTTP GET https://index.rubygems.org/info/tins
HTTP GET https://index.rubygems.org/info/backports
HTTP GET https://index.rubygems.org/info/cucumber-tag_expressions
HTTP GET https://index.rubygems.org/info/facets
HTTP GET https://index.rubygems.org/info/trollop
HTTP 200 OK https://index.rubygems.org/info/facets
HTTP 200 OK https://index.rubygems.org/info/trollop
HTTP 200 OK https://index.rubygems.org/info/cucumber-tag_expressions
HTTP 200 OK https://index.rubygems.org/info/backports
HTTP 200 OK https://index.rubygems.org/info/tins
Looking up gems ["curses"]
HTTP GET https://index.rubygems.org/info/curses
HTTP 200 OK https://index.rubygems.org/info/curses
Fetching version metadata from https://rubygems.org/
Looking up gems ["hoe", "racc", "rake-compiler", "rexical", "tenderlove-frex", "weakling", "mini_portile", "mini_portile2", "pkg-config", "rake", "rubyforge", "RubyInline", "gemcutter", "json", "net-scp", "json_pure", "ZenTest", "spruz", "net-ssh", "echoe", "needle", "jruby-pageant", "bcrypt_pbkdf", "rbnacl", "rbnacl-libsodium", "ffi", "highline", "rcov", "allison", "rdoc", "termios", "minitest", "bones", "little-plugger", "loquacious", "bones-extras", "bones-git", "rspec", "bones-rcov", "bones-rspec", "bones-rubyforge", "bones-zentest", "cucumber", "diff-lcs", "spicycode-rcov", "syntax", "rspec-core", "rspec-expectations", "rspec-mocks", "git", "rspec-support", "builder", "polyglot", "term-ansicolor", "treetop", "gherkin", "multi_json", "multi_test", "cucumber-core", "gherkin3", "cucumber-wire", "event-bus", "cucumber-expressions", "tins", "backports", "cucumber-tag_expressions", "facets", "trollop", "curses"]
Resolving dependencies...
Using bundler 1.15.4
0:  bundler (1.15.4) from C:/Ruby21/lib/ruby/gems/2.1.0/specifications/bundler-1.15.4.gemspec
Fetching mini_portile2 2.2.0
Installing mini_portile2 2.2.0
0:  mini_portile2 (2.2.0) from C:/Ruby21/lib/ruby/gems/2.1.0/specifications/mini_portile2-2.2.0.gemspec
Fetching nokogiri 1.8.0 (x86-mingw32)
Installing nokogiri 1.8.0 (x86-mingw32)
Gem::InstallError: nokogiri requires Ruby version < 2.5, >= 2.2.
C:/Ruby21/lib/ruby/site_ruby/2.1.0/rubygems/installer.rb:611:in
`ensure_required_ruby_version_met'
C:/Ruby21/lib/ruby/site_ruby/2.1.0/rubygems/installer.rb:841:in
`pre_install_checks'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/rubygems_gem_installer.rb:17:in
`pre_install_checks'
  C:/Ruby21/lib/ruby/site_ruby/2.1.0/rubygems/installer.rb:279:in `install'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/source/rubygems.rb:145:in
`block in install'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/rubygems_integration.rb:195:in
`preserve_paths'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/source/rubygems.rb:135:in
`install'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/installer/gem_installer.rb:55:in
`install'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/installer/gem_installer.rb:15:in
`install_from_spec'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/installer/parallel_installer.rb:115:in
`block in worker_pool'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:63:in
`call'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:63:in
`apply_func'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:58:in
`block in process_queue'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:55:in
`loop'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:55:in
`process_queue'
C:/Ruby21/lib/ruby/gems/2.1.0/gems/bundler-1.15.4/lib/bundler/worker.rb:89:in
`block (2 levels) in create_threads'
An error occurred while installing nokogiri (1.8.0), and Bundler cannot
continue.
Make sure that `gem install nokogiri -v '1.8.0'` succeeds before bundling.
In Gemfile:
  nokogiri
Command exited with code 5
```

**Comment by leoarnold** [comment]
BTW: The issue can be reproduced on [AppVeyor](https://ci.appveyor.com) using

```yaml
# appveyor.yml
version: '{build}'

skip_tags: true

os: Windows Server 2012

environment:
  matrix:
    - RUBY_VERSION: 21
    - RUBY_VERSION: 200
    - RUBY_VERSION: 193

install:
  - set PATH=C:\Ruby%RUBY_VERSION%\bin;%PATH%
  - gem update --system
  - gem install bundler -v 1.15.4

test_script:
  - set PATH=C:\Ruby%RUBY_VERSION%\bin;%PATH%
  - bundle config --local specific_platform true
  - bundle env
  - bundle install -V

build: off
```

**Comment by MSP-Greg** [comment]
If you have DevKit installed, the following might work.  It should stop the binary gem from being used, (although you might want to add logic in your appveyor.yml file to disable that for supported ruby versions):
```
bundle config build.nokogiri  --platform ruby
```

You make also need a `-rdevkit` in there to load the DevKit path.

From the what you've mentioned, the binary gem does not have files for Ruby previous to 2.2, so you have to force it to compile.  Currently, Nokogiri tests (and builds) against 1.93 thru 2.3, see [Appveyor](https://ci.appveyor.com/project/flavorjones/nokogiri), so it can be done.

**Comment by leoarnold** [comment]
@MSP-Greg Thus issue is not meant to be about sharing workarounds. Many Windows users told me they resent Bundler because it introduces so many problems. If I tell them to bend their systems to fit Bundler, they'd rather install all gems manually. The toy example above is just to help the Bundler team find the bug.

**Comment by MSP-Greg** [comment]
@leoarnold 

Forgive the use of an American phrase, but you're 'preaching to the choir'.  Please see https://github.com/rubygems/rubygems/commit/9103b81c6053daeafc2c74e654daffafa796b8f1

Generalizing your use case:

1. On Windows, we'd like to support ruby versions ( < 2.2) that are currently unsupported for both maintenance & security.
2. We require an extension gem that has a 'fat binary' supporting ruby 2.2, 2.3, & 2.4.
3. Bundler/rubygems won't force a compile of the gem under these conditions.
4. Please patch Bundler/rubygems to correct this.

As to my feelings about Bundler, please see #6000, and the lack of an appveyor.yml file.

'Ruby on Windows' has a long way to go.  A of people have worked on it in the past, and more people need to work on it now.  As to bundler/rubygems, I use trunk, and it's not pretty.

Summing up, given your use case, and all the entities involved,'work-a-rounds' are the only short term solution.  If you'd like to try to narrow the scope of your bug/issue, relatively current docs [here](https://msp-greg.github.io/bundler/index.html).

**Comment by segiddins** [comment]
No one on the Bundler core team uses Windows, and no one has ever submitted a PR getting CI on Windows set up. We'd welcome contributions improving the state of Windows support, though!

**Comment by MSP-Greg** [comment]
I started on it, I'm kind of bouncing between several (you may recall my RubyGems PR's).

I didn't mean to imply that I consider it a problem caused by the Bundler team.  Windows users cannot expect *nix and/or OSX users to fix Windows issues...

**Comment by segiddins** [comment]
Wow, I was totally wrong. This has nothing to do with a resolver, and has everything to do with Bundler & rubygems.org having a different understand on how to serialize `required_ruby_version` with multiple requirements in https://index.rubygems.org/info/nokogiri. Will investigate further

**Comment by segiddins** [comment]
https://github.com/rubygems/rubygems.org/pull/1669 should fix this

**Comment by MSP-Greg** [comment]
@segiddins 

Thanks for the PR.  Not knowing anything about the web site side of things, once it's accepted, is its effect immediate?

Am I correct in assuming this should allow me (a trunk user) to remove `--platform ruby` from `.gemrc`, and a gem like Nokogiri should install?

If so, I think you've just saved new Windows Ruby users a lot of headaches...

**Comment by segiddins** [comment]
It should be, but we'd have to do some manual cache invalidation server-side as well

**Comment by colby-swandale** [comment]
No one has commented on this in ages, so i'm going to assume that this issue has been resolved.
